### PR TITLE
Migrate scriptAndStyleTagProcessor to Typescript

### PR DESCRIPTION
**Pull request by jovyntls**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Part of #1913

**Overview of changes:**
Migrates scriptAndStyleTagProcessor to Typescript

**Anything you'd like to highlight/discuss:**
N/A

**Testing instructions:**
N/A

**Proposed commit message: (wrap lines at 72 characters)**
(merge commit)

---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -90,4 +91,4 @@ packages/core/src/Page/PageVueServerRenderer.js
 packages/core/src/External/External.js
 packages/core/src/External/ExternalManager.js
 
-# --- packages/core end ---
```
Comment (raysonkoh) [inline .gitignore:93]: Nit: unrelated change here

File: packages/core/src/html/scriptAndStyleTagProcessor.ts [hunk 2]
```diff
@@ -10,24 +11,23 @@ const cheerio = require('cheerio');
...
   // Do not process script/style tags that are from External
   const isExternal = userScriptsAndStyles === undefined;
   if (isHeadOrBottom || isExternal) {
     return;
   }
 
-  const idx = node.parent.children.indexOf(node);
-  if (idx !== -1) {
-    node.parent.children.splice(idx, 1);
+  const nodeSiblings = node.parent?.children;
```
Comment (raysonkoh) [inline packages/core/src/html/scriptAndStyleTagProcessor.ts:24]: Personally I would do `nodeSiblings = node.parent ?? []` as it removes the need for the if check on the next line. Not a big deal, you can leave it if you want to.

**Review by raysonkoh — APPROVED** [review]
LGTM

File: packages/core/src/html/scriptAndStyleTagProcessor.ts [hunk 3]
```diff
@@ -10,24 +11,23 @@ const cheerio = require('cheerio');
  * @param node {Object<any, any>} node from the dom traversal
```
Comment (tlylt) [inline packages/core/src/html/scriptAndStyleTagProcessor.ts:11]: Is `Object<any, any>` an accurate description of `DomElement` ?

Shall we remove the type here in JSdocs as that information is now available directly in TypeScript.

File: packages/core/src/html/scriptAndStyleTagProcessor.ts (hunk 2 above)
Comment (jovyntls) [reply packages/core/src/html/scriptAndStyleTagProcessor.ts:24]: I like that better too, thanks for the suggestion!

File: packages/core/src/html/scriptAndStyleTagProcessor.ts (hunk 3 above)
Comment (jovyntls) [reply packages/core/src/html/scriptAndStyleTagProcessor.ts:11]: Right, thanks for the catch!

**Comment by jovyntls** [comment]
@tlylt @raysonkoh Your previous comments may not be appearing on the Github diff because I've force-pushed the branch on my fork to keep the two rename/adapt commits. Do let me know if there's a better way of handling the commits for the merge later on!

File: packages/core/src/html/scriptAndStyleTagProcessor.ts [hunk 4]
```diff
@@ -7,27 +8,24 @@ const cheerio = require('cheerio');
...
     return;
   }
 
-  const idx = node.parent.children.indexOf(node);
+  const nodeSiblings = node.parent?.children ?? [node];
+  const idx = nodeSiblings.indexOf(node);
   if (idx !== -1) {
-    node.parent.children.splice(idx, 1);
+    nodeSiblings.splice(idx, 1);
   }
```
Comment (tlylt) [inline packages/core/src/html/scriptAndStyleTagProcessor.ts:28]: Want to confirm with the latest update, will `idx` ever be -1?

Comment (jovyntls) [reply packages/core/src/html/scriptAndStyleTagProcessor.ts:28]: If somehow the node is not a child of its parent, `idx` might be -1. This should likely be guaranteed by DomHandler, and I think it's unlikely that the htmlparser will have such a case. Do you think it's safe to remove this section?

Comment (tlylt) [reply packages/core/src/html/scriptAndStyleTagProcessor.ts:28]: I think it might be safe to remove because, in that strange scenario above, the resultant array of nodes may also become erroneous (to say that they are siblings of the input node).

**Comment by raysonkoh** [comment]
> Do let me know if there's a better way of handling the commits for the merge later on!

I've been using the git fixup workflow in https://github.com/MarkBind/markbind/pull/2029#issuecomment-1250189350 . If there are any changes just push a fixup commit and the reviewer just needs to review that commit. Just have to remember to squash the fixups before merging.

**Review by tlylt — APPROVED** [review]

**Review by raysonkoh — APPROVED** [review]
LGTM
